hd-wallet/src/main/java/com/bloxbean/cardano/hdwallet/Wallet.java

```diff
@@ -0,0 +1,280 @@
+        return getAccountObject(account,index).getBaseAddress();
+    }
+
+    /**
+     * Returns the Account object for the index and current account. Account can be changed via the setter.
+     * @param index
+     * @return
+     */
+    public Account getAccountObject(int index) {
+        return getAccountObject(this.account, index);
+    }
+
+    /**
+     * Returns the Account object for the index and account.
+     * @param account
+     * @param index
+     * @return
+     */
+    public Account getAccountObject(int account, int index) {
+        if(account != this.account) {
+            DerivationPath derivationPath = DerivationPath.createExternalAddressDerivationPathForAccount(account);
+            derivationPath.getIndex().setValue(index);
+            return new Account(this.network, this.mnemonic, derivationPath);
+        } else {
+            if(cache.containsKey(index)) {
+                return cache.get(index);
+            } else {
+                Account acc = new Account(this.network, this.mnemonic, index);
+                cache.put(index, acc);
+                return acc;
+            }
+        }
+    }
+
+    /**
+     * Setting the current account for derivation path.
+     * Setting the account will reset the cache.
+     * @param account
+     */
+    public void setAccount(int account) {
+        this.account = account;
+        // invalidating cache since it is only held for one account
+        cache = new HashMap<>();
+    }
+
+    /**
+     * Returns the RootkeyPair
+     * @return
+     */
+    public HdKeyPair getHDWalletKeyPair() {
+        if(rootKeys == null) {
+            HdKeyGenerator hdKeyGenerator = new HdKeyGenerator();
+            try {
+                byte[] entropy = MnemonicCode.INSTANCE.toEntropy(this.mnemonic);
+                rootKeys = hdKeyGenerator.getRootKeyPairFromEntropy(entropy);
+            } catch (MnemonicException.MnemonicLengthException | MnemonicException.MnemonicWordException |
+                     MnemonicException.MnemonicChecksumException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return rootKeys;
+    }
+
+    /**
+     * Finds needed signers within wallet and signs the transaction with each one
+     * @param txToSign
+     * @return signed Transaction
+     */
+    public Transaction sign(Transaction txToSign, WalletUtxoSupplier utxoSupplier) {
+        List<Account> signers = getSignersForTransaction(txToSign, utxoSupplier);
+
+        if(signers.isEmpty())
+            throw new RuntimeException("No signers found!");
+
+        for (Account signer : signers) txToSign = signer.sign(txToSign);
+
+        return txToSign;
+    }
+
+    /**
+     * Returns a list with signers needed for this transaction
+     *
+     * @param tx
+     * @param utxoSupplier
+     * @return
+     */
+    public List<Account> getSignersForTransaction(Transaction tx, WalletUtxoSupplier utxoSupplier) {
```

Comment:
@gavinharris-dev This is good feedback. I think instead of returning `List<Account>`, it can return `List<TxSigner>` or just `TxSigner` by composing multiple signers.

I agree that we should try to resolve all possible signers for the above scenarios. There are a few additional governance-related signers introduced in the Conway era where we need to sign using Drep credentials.

As the Wallet API will be compatible with the QuickTx API, additional signers can also be added in a standard way.